Double scrollbars and other terrible rendering on Windows

The chat seems really thin for me, making many images stretch. Could this be fatter?

![image](https://cloud.githubusercontent.com/assets/12688112/25824534/ccbc0d94-340d-11e7-9683-f87106d8ddd4.png)

Hmm... I got a bunch of errors on building

`npm WARN deprecated tough-cookie@0.12.1: ReDoS vulnerability parsing Set-Cookie https://nodesecurity.io/advisories/130
npm WARN deprecated wrench@1.5.4: wrench.js is deprecated! You should check out fs-extra (https://github.com/jprichardson/node-fs-extra) for any operations you were using wrench for. Thanks for all the usage over the years.
npm WARN deprecated minimatch@0.2.12: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated graceful-fs@1.2.3: graceful-fs v3.0.0 and before will fail on node releases >= v7.0. Please update to graceful-fs@^4.0.0 as soon as possible. Use 'npm ls graceful-fs' to find it in the tree.
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue`

and

`npm ERR! Windows_NT 10.0.10586
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install" "--no-optional"
npm ERR! node v6.5.0
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE
npm ERR! hangr@ postinstall: `cd node_modules/hangupsjs && ../.bin/coffee -o lib src`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the hangr@ postinstall script 'cd node_modules/hangupsjs && ../.bin/coffee -o lib src'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the hangr package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cd node_modules/hangupsjs && ../.bin/coffee -o lib src
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs hangr
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls hangr
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\hangr\hangr\npm-debug.log`

--- a/app/app.js
+++ b/app/app.js
@@ -3,12 +3,12 @@
 const electron = require('electron'),
     fs = require('fs-extra'),
     path = require('path'),
-    {app, BrowserWindow, ipcMain, Menu, Tray} = electron,
+    { app, BrowserWindow, ipcMain, Menu, Tray } = electron,
     // packageJson = require(__dirname + '/package.json'),
 
-    {urlForConvId} = require('./util'),
-    {IpcHandler} = require('./ipc'),
-    {dockManager, DockedWindow} = require('./docked-window'),
+    { isWindows, urlForConvId } = require('./util'),
+    { IpcHandler } = require('./ipc'),
+    { dockManager , DockedWindow} = require('./docked-window'),
     isDemo = process.argv.length && process.argv[process.argv.length - 1] === 'demo',
     ConnectionManager = isDemo
         ? require('./demo/connection')
@@ -221,8 +221,10 @@ app.on('ready', () => {
     // show the main window
     showMainWindow();
 
-    Menu.setApplicationMenu(
-        Menu.buildFromTemplate(menuTemplate));
+    if (!isWindows) {
+        Menu.setApplicationMenu(
+            Menu.buildFromTemplate(menuTemplate));
+    }
 
     if (devConfig['dev-tools'] === true) {
         mainWindow.openDevTools();

--- a/app/app.js
+++ b/app/app.js
@@ -1,21 +1,20 @@
 'use strict';
 
 const electron = require('electron'),
-      fs = require('fs-extra'),
-      path = require('path'),
-      {app, BrowserWindow, ipcMain, Menu, Tray} = electron,
-      // packageJson = require(__dirname + '/package.json'),
+    fs = require('fs-extra'),
+    path = require('path'),
+    {app, BrowserWindow, ipcMain, Menu, Tray} = electron,
+    // packageJson = require(__dirname + '/package.json'),
 
-      {urlForConvId} = require('./util'),
-      {IpcHandler} = require('./ipc'),
-      {dockManager, DockedWindow} = require('./docked-window'),
-      isDemo = process.argv.length &&
-        process.argv[process.argv.length - 1] === 'demo',
-      ConnectionManager = isDemo
+    {urlForConvId} = require('./util'),
+    {IpcHandler} = require('./ipc'),
+    {dockManager, DockedWindow} = require('./docked-window'),
+    isDemo = process.argv.length && process.argv[process.argv.length - 1] === 'demo',
+    ConnectionManager = isDemo
         ? require('./demo/connection')
         : require('./connection'),
 
-      connMan = new ConnectionManager();
+    connMan = new ConnectionManager();
 
 const devConfigFile = __dirname + '/config.json';
 var devConfig = {};
@@ -274,6 +273,8 @@ app.on('ready', () => {
         systemTray.popUpContextMenu(trayMenu);
     });
 
-    // hide the dock icon; we have a tray icon!
-    app.dock.hide();
+    if (app.dock) {
+        // hide the dock icon; we have a tray icon!
+        app.dock.hide();
+    }
 });

--- a/app/app.js
+++ b/app/app.js
@@ -215,6 +215,9 @@ ipcHandler.attach(ipcMain);
 app.on('before-quit', preQuit);
 
 app.on('ready', () => {
+    systemTray = new Tray(trayIcons.light);
+    dockManager.setTrayBounds(systemTray.getBounds());
+
     // show the main window
     showMainWindow();
 
@@ -254,7 +257,6 @@ app.on('ready', () => {
     // though, we want to be able to just click the system
     // menu icon to open the friends list
     const trayMenu = Menu.buildFromTemplate(trayContextMenu);
-    systemTray = new Tray(trayIcons.light);
     systemTray.on('click', (e) => {
         if (e.altKey) {
             systemTray.popUpContextMenu(trayMenu);

--- a/app/css/main.less
+++ b/app/css/main.less
@@ -1,6 +1,7 @@
 // theming
 
 @message-bg-incoming: #f0f0f0;
+
 @read-indicator-avatar-size: 20px;
 @video-call-accent: #1D9C0D;
 
@@ -125,9 +126,17 @@ body {
     left: 0;
     right: 0;
     bottom: 0;
-    overflow-y: scroll;
 
     padding-top: 8px;
+
+    .scroll-host {
+        overflow-y: scroll;
+        position: absolute;
+        top: 0;
+        left: 0;
+        right: 0;
+        bottom: 0;
+    }
 }
 
 #loading {

--- a/app/demo/data/recent-conversations.json
+++ b/app/demo/data/recent-conversations.json
@@ -1,4 +1,4 @@
-[
+
   {
     "conversation_id": {
       "id": "my-chat-id"

--- a/app/docked-window.js
+++ b/app/docked-window.js
@@ -1,5 +1,7 @@
 'use strict';
 
+const isWindows = require('os').platform() === 'win32';
+
 var electron;
 try {
     electron = require('electron');
@@ -7,7 +9,7 @@ try {
     // unit test mode
     electron = {};
 }
-const {BrowserWindow} = electron;
+const { BrowserWindow } = electron;
 
 const WindowDimens = {
     w: 240,
@@ -41,6 +43,10 @@ class DockManager {
     constructor() {
         this._anchor = Anchor.End;
         this._windows = [];
+        this._trayOffset = {
+            x: 0,
+            y: 0,
+        };
     }
 
     adopt(window) {
@@ -99,7 +105,7 @@ class DockManager {
     position(window, animate) {
         var screenHeight = screenSize().height;
         var x = this._anchor.position(window._index);
-        var y = screenHeight - window.height;
+        var y = screenHeight - window.height + this._trayOffset.y;
 
         // special case: when animating in, we like a nice slide
         if (animate === 'in') {
@@ -111,6 +117,20 @@ class DockManager {
         }
 
         window.setPosition(x, y, !!animate);
+    }
+
+    setTrayBounds(bounds) {
+        if (isWindows) {
+            // NOTE: we currently only care about these on Windows
+            return;
+        }
+
+        const { height } = screenSize();
+        if (bounds.y === height - bounds.height) {
+            // system tray is on the bottom; move our
+            // windows up above it
+            this._trayOffset.y = -bounds.height;
+        }
     }
 
     /**
@@ -188,6 +208,12 @@ class DockedWindow {
             skipTaskBar: true,
         });
         win.__docked = this;
+
+        if (isWindows) {
+            // NOTE: hiding the menu on windows to reduce visual clutter
+            // is this the right way to approach this?
+            win.setMenu(null);
+        }
 
         // and load the index.html of the app.
         win.loadURL('file://' + __dirname + '/index.html#' + url);

--- a/app/docked-window.js
+++ b/app/docked-window.js
@@ -1,7 +1,5 @@
 'use strict';
 
-const isWindows = require('os').platform() === 'win32';
-
 var electron;
 try {
     electron = require('electron');
@@ -10,6 +8,8 @@ try {
     electron = {};
 }
 const { BrowserWindow } = electron;
+
+const { isWindows } = require('./util');
 
 const WindowDimens = {
     w: 240,

--- a/app/docked-window.js
+++ b/app/docked-window.js
@@ -120,7 +120,7 @@ class DockManager {
     }
 
     setTrayBounds(bounds) {
-        if (isWindows) {
+        if (!isWindows) {
             // NOTE: we currently only care about these on Windows
             return;
         }

--- a/app/util.js
+++ b/app/util.js
@@ -27,7 +27,7 @@ module.exports.throttle = function throttle(fn, periodMs) {
 
     function throttled(...args) {
         if (state.token) clearTimeout(state.token);
-        
+
         var now = Date.now();
         var leftInPeriod = periodMs - (now - state.periodStart);
         if (leftInPeriod <= 0) {

--- a/app/util.js
+++ b/app/util.js
@@ -1,4 +1,6 @@
 
+module.exports.isWindows = require('os').platform() === 'win32';
+
 module.exports.urlForConvId = function urlForConvId(convId) {
     return `/c/${convId}`;
 };

--- a/src/cljs/hangr/views/friends_list.cljs
+++ b/src/cljs/hangr/views/friends_list.cljs
@@ -99,9 +99,10 @@
       (let [convs @convs]
         (if (seq convs)
           ;; we have conversations
-          [:ul#conversations
-           (for [c convs] 
-             ^{:key (:id c)} [friends-list-item c])]
+          [:div.scroll-host
+           [:ul#conversations
+            (for [c convs]
+              ^{:key (:id c)} [friends-list-item c])]]
           ;; nothing :(
           [:div "No conversations"])))))
 

